workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/DisplayRendering.kt

```diff
@@ -0,0 +1,233 @@
+ *
+ * Sets the workflow rendering associated with this [View], and displays it by
+ * invoking the [DisplayRendering] function previously set by [View.bindShowRendering].
+ *
+ * @throws IllegalStateException if [View.bindShowRendering] has not been called.
+ */
+@WorkflowUiExperimentalApi
+fun <RenderingT : ViewRendering> View.display(
+  rendering: RenderingT,
+  viewEnvironment: ViewEnvironment
+) {
+  showRenderingTag
+      ?.let { tag ->
+        check(tag.showing.matches(rendering)) {
+          "Expected $this to be able to show rendering $rendering, but that did not match " +
+              "previous rendering ${tag.showing}. " +
+              "Consider using ${WorkflowViewStub::class.java.simpleName} to display arbitrary types."
+        }
+
+        bindShowRendering(rendering, viewEnvironment, tag.displayRendering)
+      }
+      ?: error(
+          "Expected $this to have a showRendering function to show $rendering. " +
+              "Perhaps it was not built by a ${ViewBuilder::class.java.simpleName}, " +
+              "or perhaps its ${ViewBuilder::class.java.simpleName} did not call" +
+              "View.bindShowRendering."
+      )
+}
+
+/**
+ * Sets the workflow rendering associated with this [Dialog], and displays it by
+ * invoking the [DisplayRendering] function previously set by [Dialog.bindDisplayFunction].
+ *
+ * @throws IllegalStateException if [Dialog.bindDisplayFunction] has not been called.
+ */
+@WorkflowUiExperimentalApi
+fun <RenderingT : ModalRendering> Dialog.display(
+  rendering: RenderingT,
+  viewEnvironment: ViewEnvironment
+) {
+  window!!.decorView.showRenderingTag
+      ?.let { tag ->
+        check(tag.showing.matches(rendering)) {
+          "Expected $this to be able to display rendering $rendering, but that did not match " +
+              "previous rendering ${tag.showing}. "
+        }
+
+        bindDisplayFunction(rendering, viewEnvironment, tag.displayRendering)
+      }
+      ?: error(
+          "Expected $this to have a showRendering function to display $rendering. " +
+              "Perhaps it was not built by a ${DialogBuilder::class.java.simpleName}, " +
+              "or perhaps its ${DialogBuilder::class.java.simpleName} did not call" +
+              "Dialog.bindShowRendering."
+      )
+}
+
+/**
+ * Returns the most recent rendering displayed by this [View], or null if [View.bindShowRendering]
+ * has never been called.
+ */
+@WorkflowUiExperimentalApi
+fun <RenderingT : ViewRendering> View.getDisplaying(): RenderingT? {
+  // Can't use a val because of the parameter type.
+  @Suppress("UNCHECKED_CAST")
+  return when (val displaying = showRenderingTag?.showing) {
+    null -> null
+    else -> displaying as RenderingT
+  }
+}
+
+/**
+ * Returns the most recent rendering displayed by this [Dialog], or null if [Dialog.bindDisplayFunction]
+ * has never been called.
+ */
+@WorkflowUiExperimentalApi
+fun <RenderingT : ModalRendering> Dialog.getDisplaying(): RenderingT? {
```

Comment:
Is it worth an interface with `showRenderingTag` and `displaying` so we can abstract some of these duplicated functions into the same implementation?

Reply:
nvm I see these extension properties are defined below (which of course they are as this is Dialog and View). No clear way to pull out an abstraction as an extension on Dialog and View I guess. 🤔

Reply:
I guess you do almost already have an extension abstraction in ShowRenderingTag - why can't all these other extensions delegate into functions defined on that class itself. That way we wouldn't have the double maintenance. Granted this may not scale much beyond Dialog and View.

Reply:
Except for would it if we introduce Compose?